internal/api/mail.go

```diff
@@ -575,34 +525,71 @@ func (a *API) sendEmail(r *http.Request, tx *storage.Connection, u *models.User,
 			EmailActionType: emailActionType,
 			RedirectTo:      referrerURL,
 			SiteURL:         externalURL.String(),
-			TokenHash:       tokenHashWithPrefix,
-		}
-		if emailActionType == mail.EmailChangeVerification && config.Mailer.SecureEmailChangeEnabled && u.GetEmail() != "" {
-			emailData.TokenNew = otpNew
-			emailData.TokenHashNew = u.EmailChangeTokenCurrent
+			TokenHash:       ott.TokenHash,
 		}
 		input := hooks.SendEmailInput{
 			User:      u,
 			EmailData: emailData,
 		}
 		output := hooks.SendEmailOutput{}
-		return a.invokeHook(tx, r, &input, &output, a.config.Hook.SendEmail.URI)
+		return a.invokeHook(tx, r, &input, &output, config.Hook.SendEmail.URI)
 	}
 
 	switch emailActionType {
 	case mail.SignupVerification:
-		return mailer.ConfirmationMail(r, u, otp, referrerURL, externalURL)
+		return mailer.ConfirmationMail(r, u, ott, otp, referrerURL, externalURL)
 	case mail.MagicLinkVerification:
-		return mailer.MagicLinkMail(r, u, otp, referrerURL, externalURL)
+		return mailer.MagicLinkMail(r, u, ott, otp, referrerURL, externalURL)
 	case mail.ReauthenticationVerification:
 		return mailer.ReauthenticateMail(r, u, otp)
 	case mail.RecoveryVerification:
-		return mailer.RecoveryMail(r, u, otp, referrerURL, externalURL)
+		return mailer.RecoveryMail(r, u, ott, otp, referrerURL, externalURL)
 	case mail.InviteVerification:
-		return mailer.InviteMail(r, u, otp, referrerURL, externalURL)
-	case mail.EmailChangeVerification:
-		return mailer.EmailChangeMail(r, u, otpNew, otp, referrerURL, externalURL)
+		return mailer.InviteMail(r, u, ott, otp, referrerURL, externalURL)
 	default:
 		return errors.New("invalid email action type")
 	}
 }
+
+func (a *API) sendEmailChangeEmails(r *http.Request, tx *storage.Connection, u *models.User, ottNew, ottCurrent *models.OneTimeToken, emailActionType, otp, otpNew string) error {
+	mailer := a.Mailer()
+	ctx := r.Context()
+	config := a.config
+	referrerURL := utilities.GetReferrer(r, config)
+	externalURL := getExternalHost(ctx)
+	if config.Hook.SendEmail.Enabled {
+		emailData := mail.EmailData{
+			Token:           otpNew,
+			EmailActionType: emailActionType,
+			RedirectTo:      referrerURL,
+			SiteURL:         externalURL.String(),
+			TokenHash:       ottNew.TokenHash,
+		}
+		if config.Mailer.SecureEmailChangeEnabled && u.GetEmail() != "" {
+			emailData.TokenNew = otp
+			emailData.TokenHashNew = ottCurrent.TokenHash
+		}
+		input := hooks.SendEmailInput{
+			User:      u,
+			EmailData: emailData,
+		}
+		output := hooks.SendEmailOutput{}
+		return a.invokeHook(tx, r, &input, &output, config.Hook.SendEmail.URI)
+	}
+
+	errors := make(chan error)
+	if config.Mailer.SecureEmailChangeEnabled && u.GetEmail() != "" {
+		go func(c chan error) {
+			c <- mailer.EmailChangeMail(r, u, ottCurrent, otp, referrerURL, externalURL)
+		}(errors)
+	}
+	go func(c chan error) {
+		c <- mailer.EmailChangeMail(r, u, ottNew, otpNew, referrerURL, externalURL)
+	}(errors)
```

Comment:
previously, `EmailChangeMail` handled multiplexing between the current and new email - i find it alot cleaner for it just to do 1 thing (send the email for email change) and have a separate function that wraps it to decide whether to send 2 emails or just 1